gk/main.c

```diff
@@ -1015,10 +1143,13 @@ process_pkts_front(uint8_t port_front, uint8_t port_back,
 
 	/* XXX Do something better here! For now, free any unsent packets. */
 	if (unlikely(num_tx_succ < num_tx)) {
-		for (i = num_tx_succ; i < num_tx; i++)
+		for (i = num_tx_succ; i < num_tx; i++) {
+			tx_bytes -= tx_bufs[i]->pkt_len;
 			drop_packet(tx_bufs[i]);
+		}
 	}
 
+	rte_red_mark_queue_empty(&instance->red, tx_bytes);
```

Comment:
Calling `rte_red_mark_queue_empty()` here is really weird because it means that the queue won't ever go above any meaningful threshold. To see this, assume that the RED queue of a given GK block is empty, some packets arrive, the GK block does a number of calls to `red_check()`, and then calls `rte_red_mark_queue_empty()` to state that the queue is empty. The only way that the RED queue will be above the threshold is if the threshold is smaller than the number of packets that a GK block reads from the NICs at once.

The issue is that GK blocks don't really have a queue before transmitting the packets. While we could implement this queue, I think we don't want it because it would overflow the cache of the processors and slow Gatekeeper down. We are already taking this risk with the Solicitator block, but it only drives 5% of the bandwidth.

We need to reflect on this issue, but, right now, it seems that Gatekeeper servers must forgo the RED queues and leave the job to the routers on the path from the Gatekeeper server to Grantor servers. Notice that Gatekeeper servers already don't implement the hashed, granted RED queues, so it's not like we are given more work to routers than what we had previously given.